eng/scripts/Verify-And-Merge-PRs.ps1

```diff
@@ -22,43 +23,54 @@ if ($AuthToken) {
 
 $PRData = Get-Content $PRDataArtifactPath
 # Confirm Mergability
-foreach ($prDataLine in $PRData)
-{
-  $repoOwner, $repoName, $prNumber = $prDataLine.Split(";")
 
-  $prApiUrl = "https://api.github.com/repos/${repoOwner}/${repoName}/pulls/${prNumber}"
-  $prUrl = "https://github.com/${repoOwner}/${repoName}/pull/${prNumber}"
-  
-  try
+do 
+{
+  $unMergablePRs = @()
+  foreach ($prDataLine in $PRData)
   {
-    $response = Invoke-RestMethod -Headers $headers $prApiUrl
-    if ($response.merged) {
-      Write-Host "${prUrl} is merged."
-    }
-    elseif ($response.state -eq "closed") {
-      LogWarning "${prUrl} is closed. Please investigate why was not merged."
-      $ReadyForMerge = $false
-    }
-    elseif ($response.mergeable -and $response.mergeable_state -eq "clean") {
-      Write-Host "${prUrl} is ready to merge."
+    $repoOwner, $repoName, $prNumber = $prDataLine.Split(";")
 
-      $mergablePRs += @{ Url = $prApiUrl; HeadSHA = $response.head.sha }
-    }
-    elseif ($response.mergeable_state -ne "clean") {
-      LogWarning "${prUrl} is blocked ($($response.mergeable_state)). Please ensure all checks are green and reviewers have approved."
-      $ReadyForMerge = $false
+    $prApiUrl = "https://api.github.com/repos/${repoOwner}/${repoName}/pulls/${prNumber}"
+    $prUrl = "https://github.com/${repoOwner}/${repoName}/pull/${prNumber}"
+    
+    try
+    {
+      $response = Invoke-RestMethod -Headers $headers $prApiUrl
+      if ($response.merged) {
+        Write-Host "${prUrl} is merged."
+      }
+      elseif ($response.state -eq "closed") {
+        LogWarning "${prUrl} is closed. Please investigate why was not merged."
+        $ReadyForMerge = $false
+      }
+      elseif ($response.mergeable -and $response.mergeable_state -eq "clean") {
+        Write-Host "${prUrl} is ready to merge."
+
+        $mergablePRs += @{ Url = $prApiUrl; HeadSHA = $response.head.sha }
+      }
+      elseif ($response.mergeable_state -ne "clean") {
+        LogWarning "${prUrl} is blocked ($($response.mergeable_state)). Please ensure all checks are green and reviewers have approved."
+        $ReadyForMerge = $false
+        $unMergablePRs += $prDataLine
+      }
+      else {
+        LogWarning "${prUrl} is in an unknown state please contact engineering system team to understand the state."
+        LogWarning $response
+        $ReadyForMerge = $false
+        $unMergablePRs += $prDataLine
+      }
     }
-    else {
-      LogWarning "${prUrl} is in an unknown state please contact engineering system team to understand the state."
-      LogWarning $response
-      $ReadyForMerge = $false
+    catch {
+      LogError "Invoke-RestMethod ${prApiUrl} failed with exception:`n$_"
+      exit 1
     }
   }
-  catch {
-    LogError "Invoke-RestMethod ${prApiUrl} failed with exception:`n$_"
-    exit 1
-  }
+  $PRData = $unMergablePRs
+  $RetryCount--
 }
+while (!$ReadyForMerge -and ($RetryCount -gt 0) -and $ShouldMerge)
```

Comment:
We should use the $PRData.Count instead of ReadyForMerge because that is set to false for other reasons as well. The only reason we should loop again is if we have more things to process.